### PR TITLE
Sparse fill empty rows op

### PR DESCRIPTION
**Pull request by codeislife99**
This PR is to support SparseFillEmptyRows Op (similar to https://www.tensorflow.org/api_docs/python/tf/sparse/fill_empty_rows) in relay. Frontend code for the TF Op will be added in a following PR since it combines multiple ops together)

### CODE REVIEW

**Comment by codeislife99** [comment]
cc : @trevor-m  @comaniac @zhiics

File: python/tvm/relay/op/_transform.py
Comment (tkonolige) [inline python/tvm/relay/op/_transform.py:66]: I think `sparse_fill_empty_rows` fits better with our current naming convention.

File: python/tvm/relay/op/transform.py
Comment (tkonolige) [inline python/tvm/relay/op/transform.py:1390]: I'm not sure this example is correct. `[-1, -1]` is not a valid sparse index. Shouldn't the output be
```
new_sparse_indices=[[0, 1],
                                      [0, 3]
                                      [1, 0]
                                      [2, 0]
                                      [3, 1]
                                      [4, 0]]

new_sparse_values = [1, 2, 10, 3, 4, 10]
```

File: python/tvm/relay/op/transform.py
Comment (tkonolige) [inline python/tvm/relay/op/transform.py:1351]: What is `to_be_discarded` it is not mentioned anywhere?

File: src/relay/op/tensor/transform.cc
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1567]:
```suggestion
  ICHECK(param != nullptr);
```

File: src/relay/op/tensor/transform.cc
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1561]: Could you add an error message here. Something like "SparseFillEmptyRowsRel expects 4 arguments, but X were provided".

File: src/relay/op/tensor/transform.cc
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1584]: Can you add an error message here.

File: src/relay/op/tensor/transform.cc
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1586]:
```suggestion
  ICHECK(param != nullptr);
```

File: src/relay/op/tensor/transform.cc
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1601]: Could you update the description.

**Review by tkonolige — CHANGES_REQUESTED** [review]
Overall this looks good to me, but there are some minor changes around documentation/naming.

File: python/tvm/relay/op/transform.py
Comment (codeislife99) [reply python/tvm/relay/op/transform.py:1390]: So , if you look at the description (`N + dense_shape[0]` ) <-- this is the dim0 shape of the tensor. This is because we can't know beforehand how many rows are missing. This operation would be succeeded by strided_slice to remove with begin as 0 and end as `slice_element_index` to achieve the desired output

Comment (codeislife99) [reply python/tvm/relay/op/transform.py:1390]: Let me know if you have any better suggestions for implementation, else I will stick with this.

Comment (tkonolige) [reply python/tvm/relay/op/transform.py:1390]: Ok, I understand now. Could you add a little more documentation to that effect? I would also fill the values associated invalid indices with zeros instead of -1. That way even if someone uses the indices, the effect would be nothing.

Also, have you considered using dynamic shape instead? That would avoid the need for the negative indices.

File: python/tvm/relay/op/_transform.py
Comment (codeislife99) [reply python/tvm/relay/op/_transform.py]: Changed.

File: python/tvm/relay/op/transform.py
Comment (codeislife99) [reply python/tvm/relay/op/transform.py:1390]: No I haven't considered that. I noticed that for input dependent ops like `non_maximum_suppression`, they used nms with fixed size + strided slice and I just borrowed that idea. Do you mind pointing me to an implementation similar to what you are talking about ?

Comment (mbrookhart) [reply python/tvm/relay/op/transform.py:1390]: We have it set up that way because that's how MXNet does it (TOPI's original opset was mostly cloned from MXNet), and MXNet is fairly conservative about static shapes. From a performance perspective, if the expectation is that we'll always use this with a dynamic strided slice after, I'd think it would be better to just emit a dynamic shape. If we have another usecase, however, I think this is fine.

Do you have a framework level usecase in mind?

Comment (codeislife99) [reply python/tvm/relay/op/transform.py:1390]: Yeah, in the short-term it will be used for TF. https://www.tensorflow.org/api_docs/python/tf/sparse/fill_empty_rows

File: src/relay/op/tensor/transform.cc
Comment (codeislife99) [reply src/relay/op/tensor/transform.cc]: Done

File: src/relay/op/tensor/transform.cc
Comment (codeislife99) [reply src/relay/op/tensor/transform.cc]: Done

File: src/relay/op/tensor/transform.cc
Comment (codeislife99) [reply src/relay/op/tensor/transform.cc]: Done.

File: src/relay/op/tensor/transform.cc
Comment (codeislife99) [reply src/relay/op/tensor/transform.cc]: Done

File: src/relay/op/tensor/transform.cc
Comment (codeislife99) [reply src/relay/op/tensor/transform.cc]: Done

File: python/tvm/relay/op/transform.py
Comment (codeislife99) [reply python/tvm/relay/op/transform.py]: Changed.

**Comment by comaniac** [comment]
@tkonolige @mbrookhart PTAL and approve or request changes explicitly. Thanks.

File: include/tvm/topi/transform.h
Comment (tkonolige) [inline include/tvm/topi/transform.h:1390]:
```suggestion
 * \brief Fill empty rows of a sparse tensor with default values
```

File: python/tvm/relay/op/transform.py
Comment (tkonolige) [inline python/tvm/relay/op/transform.py:1359]: Could you update the wording here? `to_be_discarded` refers to nothing.

File: src/relay/op/tensor/transform.cc
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1567]:
```suggestion
  ICHECK_NOTNULL(param);
```

File: src/relay/op/tensor/transform.cc
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1562]:
```suggestion
  ICHECK_EQ(types.size(), 4) << "SparseFillEmptyRowsRel expects 4 arguments but " << types.size() << " were provided.";
```

File: src/relay/op/tensor/transform.cc
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1587]:
```suggestion
  ICHECK_NOTNULL(param);
```

File: src/relay/op/tensor/transform.cc
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1585]:
```suggestion
  ICHECK_EQ(inputs.size(), 3) << "SparseFillEmptyRowsCompute expects 3 arguments but " << inputs.size() << " were provided.";
```

File: src/relay/op/tensor/transform.cc
Comment (tkonolige) [inline src/relay/op/tensor/transform.cc:1607]: Please put more useful descriptions on these arguments.

**Review by tkonolige — CHANGES_REQUESTED** [review]
I've suggested some more small changes.

Given the discussion with @mbrookhart, I think you should use dynamic shape here.

**Comment by tkonolige** [comment]
@comaniac Every time I select request changes on GitHub, it switches it "suggested changes". Maybe because I am not a committer?

File: python/tvm/relay/op/transform.py
Comment (tkonolige) [reply python/tvm/relay/op/transform.py]: Given you are using it for a single op, I would follow NMS's example and use dynamic shape.

**Comment by comaniac** [comment]
> @comaniac Every time I select request changes on GitHub, it switches it "suggested changes". Maybe because I am not a committer?

Yes that's the case, but we'll do our best to make sure all comments are addressed before merging.

File: include/tvm/topi/transform.h
Comment (mbrookhart) [inline include/tvm/topi/transform.h:1415]: I would like to see this done in a way that supports dynamic shapes, this will crash a compile time if we have a dynamic input.

**Review by mbrookhart — COMMENTED** [review]
I don't fully understand the point of these sparse op PRs, is there a particular model we're trying to target? It's hard for me to review without understanding the use case.

There are issues in this set of PRs that will prevent them from working on dynamically shaped input tensors. I don't think that's a blocker, but it could cause headaches down the road.

File: include/tvm/topi/transform.h
Comment (codeislife99) [reply include/tvm/topi/transform.h:1415]: Yes there are 3 sparse ops that we are trying to target for a customer for a TF model explicitly. 
These 3 are : 
1. [sparse_reshape](https://www.tensorflow.org/api_docs/python/tf/sparse/reshape) 
2. [sparse_segment_sum](https://www.tensorflow.org/api_docs/python/tf/sparse/segment_sum?hl=bn)
3. [sparse_fill_empty_rows](https://www.tensorflow.org/api_docs/python/tf/sparse/fill_empty_rows)

Comment (codeislife99) [reply include/tvm/topi/transform.h:1415]: I am a beginner in op implementation, do you mind sharing some PR/code examples , on how I can achieve the same goal(loop etc.) and set the output shape in `OpRel`. 
Then I will follow them to make this compatible for dynamically shaped input tensors. 
There is also probably no need for the strided_slice to be outside the op implementation if I change it to support dynamic shapes.

File: python/tvm/relay/op/transform.py
Comment (codeislife99) [reply python/tvm/relay/op/transform.py]: Resolving this conversation since based on our offline discussion, we will start off with static shape

File: python/tvm/relay/op/transform.py
Comment (codeislife99) [reply python/tvm/relay/op/transform.py]: Done.

File: src/relay/op/tensor/transform.cc
Comment (codeislife99) [reply src/relay/op/tensor/transform.cc]: Done.

**Comment by codeislife99** [comment]
@tkonolige @mbrookhart This PR is ready for re-review as well whenever you are back from vacation.

**Comment by codeislife99** [comment]
Hoping back onto this stream of work after a while, was involved with something else. 
Since this PR had a lot of issues, it has been completely rewritten in #7442 and addresses all the existing comments in this PR. This Op has also been tested E2E with the customer model. Please review #7442.

**Pull request closed without merging**
